File: assignments/scala/bob/src/test/scala/bob_test.scala
import org.scalatest._

class BobSpecs extends FlatSpec with Matchers {
  def teenager = new Bob

  it should "respond to a statement" in {
    val response = teenager.hey("Tom-ay-to, tom-aaaah-to.")
    response should be ("Whatever.")
  }

  it should "respond to shouting" in {
    pending
    val response = teenager.hey("WATCH OUT!")
    response should be ("Woah, chill out!")
  }

  it should "respond to questions" in {
    pending
    val response = teenager.hey("Does this cryogenic chamber make me look fat?")
    response should be ("Sure.")
  }

  it should "allow questions to end with numbers" in {
    pending
    val response = teenager.hey("You are what, like 15?")
    response should be ("Sure.")
  }

  it should "respond to talking forcefully" in {
    pending
    val response = teenager.hey("Let's go make out behind the gym!")
    response should be ("Whatever.")
  }

  it should "allow acroynms in regular speech" in {
    pending
    val response = teenager.hey("It's OK if you don't want to go to the DMV.")
    response should be ("Whatever.")
  }

  it should "see forceful questions as shouting" in {
    pending
    val response = teenager.hey("WHAT THE HELL WERE YOU THINKING?")
    response should be ("Woah, chill out!")
  }

  it should "allow numbers when shouting" in {
    pending
    val response = teenager.hey("1, 2, 3, GO!")
    response should be ("Woah, chill out!")
  }

  it should "see only numbers as speech" in {
    pending
    val response = teenager.hey("1, 2, 3")
    response should be ("Whatever.")
  }

  it should "respond to questions with only numbers" in {
    pending
    val response = teenager.hey("4?")
    response should be ("Sure.")
  }

  it should "respond to shouting with no exclamation mark" in {
    pending
    val response = teenager.hey("I HATE YOU")
    response should be ("Woah, chill out!")
  }

  it should "respond to statements with ? in the middle" in {
    pending
    val response = teenager.hey("Ending with ? means a question.")
    response should be ("Whatever.")
  }

  it should "respond to prattling on" in {
    pending
    val response = teenager.hey("Wait! Hang on. Are you going to be OK?")
    response should be ("Sure.")
  }

  it should "respond to silence" in {
    pending
    val response = teenager.hey("")
    response should be ("Fine. Be that way!")
  }

  it should "respond to prolonged silence" in {
    pending
    val response = teenager.hey("       ")
    response should be ("Fine. Be that way!")
  }

  it should "respond to multiple line questions" in {
    pending
    val response = teenager.hey("""
Does this cryogenic chamber make me look fat?
no""")
    response should be ("Whatever.")
  }
}

File: assignments/java/bob/src/test/java/BobTest.java
import org.junit.Test;

import static org.junit.Assert.*;

public class BobTest {
    private final Bob bob = new Bob();

    @Test
    public void saySomething() {
        assertEquals(
            "Whatever.",
            bob.hey("Tom-ay-to, tom-aaaah-to.")
        );
    }

    @Test
    public void shouting() {
        assertEquals(
            "Woah, chill out!",
            bob.hey("WATCH OUT!")
        );
    }

    @Test
    public void askingAQuestion() {
        assertEquals(
            "Sure.",
            bob.hey("Does this cryogenic chamber make me look fat?")
        );
    }

    @Test
    public void askingANumericQuestion() {
        assertEquals(
            "Sure.",
            bob.hey("You are, what, like 15?")
        );
    }

    @Test
    public void talkingForcefully() {
        assertEquals(
            "Whatever.",
            bob.hey("Let's go make out behind the gym!")
        );
    }

    @Test
    public void usingAcronymsInRegularSpeech() {
        assertEquals(
            "Whatever.", bob.hey("It's OK if you don't want to go to the DMV.")
        );
    }

    @Test
    public void forcefulQuestions() {
        assertEquals(
            "Woah, chill out!", bob.hey("WHAT THE HELL WERE YOU THINKING?")
        );
    }

    @Test
    public void shoutingNumbers() {
        assertEquals(
            "Woah, chill out!", bob.hey("1, 2, 3 GO!")
        );
    }

    @Test
    public void onlyNumbers() {
        assertEquals(
            "Whatever.", bob.hey("1, 2, 3")
        );
    }

    @Test
    public void questionWithOnlyNumbers() {
        assertEquals(
            "Sure.", bob.hey("4?")
        );
    }

    @Test
    public void shoutingWithSpecialCharacters() {
        assertEquals(
            "Woah, chill out!", bob.hey("ZOMG THE %^*@#$(*^ ZOMBIES ARE COMING!!11!!1!")
        );
    }

    @Test
    public void shoutingWithUmlauts() {
        assertEquals(
            "Woah, chill out!", bob.hey("\u00dcML\u00c4\u00dcTS!")
        );
    }

    @Test
    public void calmlySpeakingWithUmlauts() {
        assertEquals(
            "Whatever.", bob.hey("\u00dcML\u00e4\u00dcTS!")
        );
    }

    @Test
    public void shoutingWithNoExclamationMark() {
        assertEquals(
            "Woah, chill out!", bob.hey("I HATE YOU")
        );
    }

    @Test
    public void statementContainingQuestionMark() {
        assertEquals(
            "Whatever.", bob.hey("Ending with ? means a question.")
        );
    }

    @Test
    public void prattlingOn() {
        assertEquals(
            "Sure.", bob.hey("Wait! Hang on. Are you going to be OK?")
        );
    }

    @Test
    public void silence() {
        assertEquals(
            "Fine. Be that way!", bob.hey("")
        );
    }

    @Test
    public void prolongedSilence() {
        assertEquals(
            "Fine. Be that way!", bob.hey("    ")
        );
    }
}
